relayer/src/worker/connection.rs

```diff
@@ -0,0 +1,127 @@
+
+        // Flag that indicates if the worker should actively resume handshake.
+        // Set on start or when event based handshake fails.
+        let mut resume_handshake = true;
+
+        loop {
+            thread::sleep(Duration::from_millis(200));
+
+            if let Ok(cmd) = self.cmd_rx.try_recv() {
+                let result = match cmd {
+                    WorkerCmd::IbcEvents { batch } => {
+                        // there can be up to two event for this connection, e.g. init and try.
+                        // process the last event, the one with highest "rank".
+                        let last_event = batch.events.last();
+
+                        debug!("connection worker starts processing {:#?}", last_event);
+
+                        match last_event {
+                            Some(event) => {
+                                let mut handshake_connection = RelayConnection::restore_from_event(
+                                    a_chain.clone(),
+                                    b_chain.clone(),
+                                    event.clone(),
+                                )?;
+
+                                retry_with_index(
+                                    retry_strategy::worker_default_strategy(),
+                                    |index| handshake_connection.step_event(event.clone(), index),
+                                )
+                            }
+                            None => Ok(()),
+                        }
+                    }
+                    WorkerCmd::NewBlock {
+                        height: current_height,
+                        new_block: _,
+                    } => {
+                        if !resume_handshake {
+                            continue;
+                        }
+
+                        debug!(
+                            "connection worker starts processing block event at {}",
+                            current_height
+                        );
+
+                        let height = current_height.decrement()?;
+
+                        let (mut handshake_connection, state) =
+                            RelayConnection::restore_from_state(
+                                a_chain.clone(),
+                                b_chain.clone(),
+                                self.connection.clone(),
+                                height,
+                            )?;
+
+                        retry_with_index(retry_strategy::worker_default_strategy(), |index| {
+                            handshake_connection.step_state(state, index)
+                        })
+                    }
+                };
+
+                if let Err(retries) = result {
+                    warn!("connection worker failed after {} retries", retries);
+
+                    // Resume handshake on next iteration.
+                    resume_handshake = true;
+                } else {
+                    resume_handshake = false;
```

Comment:
From what I can see, in the smooth case that no unexpected errors happen, then once `resume_handshake` is set to false, then it is not going to be set to true again ever. So does this means that the branch for `WorkerCmd::NewBlock` is going to always be skipped? Otherwise, are we using some subtle error variants to trigger the `Err` branch to set `resume_handshake` to true again?

Reply:
> So does this means that the branch for WorkerCmd::NewBlock is going to always be skipped? 

It's going to be executed if `step_event()` or `step_state()` have not advanced the handshake after a number of tries.